sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/PlansTracker.java

```diff
@@ -0,0 +1,297 @@
+
+    //Ensure we found our phase.
+    if (phaseList.size() != 1) {
+      JSONObject outcome = new JSONObject();
+      outcome.put("invalid_input",
+          "Supplied phase not found in set of possible phases with supplied plan!");
+      return Optional.of(outcome);
+    }
+
+    //If no explicit step defined, nothing further to do.
+    if (filterStep == null) {
+      return Optional.empty();
+    }
+
+    Phase phase = phaseList.get(0);
+    List<Step> stepList = phase.getChildren().stream()
+        .filter(steps -> steps.getName().equalsIgnoreCase(filterStep))
+        .collect(Collectors.toList());
+
+    //Ensure we found our phase.
+    if (stepList.size() != 1) {
+      JSONObject outcome = new JSONObject();
+      outcome.put("invalid_input",
+          "Supplied step not found in set of possible steps with supplied plan and phase!");
+      return Optional.of(outcome);
+    }
+
+    //Successfully found plan, phase and step.
+    return Optional.empty();
+  }
+
+  public Response getJson(@QueryParam("plan") String filterPlan,
+                        @QueryParam("phase") String filterPhase,
+                        @QueryParam("step") String filterStep,
+                        @QueryParam("sync") boolean requireSync)
+  {
+
+    //Validate plan/phase/step if provided.
+    if (filterPlan != null || filterPhase != null || filterStep != null) {
+      Optional<JSONObject> validationOutcome = getValidationErrorResponse(filterPlan,
+                                                                          filterPhase,
+                                                                          filterStep);
+      if (validationOutcome.isPresent()) {
+        return ResponseUtils.jsonOkResponse(validationOutcome.get());
+      }
+    }
+
+    //At this point we're either returning the entire plans tree or
+    //pruning it down to a plan/phase/step which has been validated.
+
+    JSONObject response = new JSONObject();
+    response.put("service-topology", generateServiceStatus(filterPlan,
+        filterPhase,
+        filterStep,
+        false,
+        true));
+
+    response.put("service-status", generateServiceStatus(filterPlan,
+        filterPhase,
+        filterStep,
+        true,
+        false));
+
+    //Retrieve the latest Plans from PlanCoordinator. Some plans i.e recovery do change over time.
+    HashMap<String, Plan> planMap = new HashMap<String, Plan>();
+    //Add all the plans for easy lookups below.
+    for (PlanManager planManager : planCoordinator.getPlanManagers()) {
+      Plan plan = planManager.getPlan();
+      planMap.put(plan.getName(), plan);
+    }
+
+    //Set scheduler to one of DEPLOYING, DEPLOYED, RECOVERING or DECOMISSIONING.
+    //If unknown, set to generic status to RUNNING when running custom-plans.
+    String schedulerStatus = "RUNNING";
+    if (planMap.containsKey(Constants.DEPLOY_PLAN_NAME) &&
+        planMap.get(Constants.DEPLOY_PLAN_NAME).isRunning())
+    {
```

Comment:
this style is inconsistent `{` again... it's different all over the file.

Reply:
This is a [rule](http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyOption.html#NLOW) enforced by checkstyle that we use